src/js/helpers/timeline/declarationHelper.js

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2020 Knot.x Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { DataSet } from 'vis-timeline';
+
+const RADIX = 10;
+const EMPTY_LABEL = '';
+
+const hasSubtasks = (node) => !!node.subtasks;
+
+const getNextNodes = (node) => node.on[node.response.transition];
+
+const hasProcessedTransition = (node) => !!node.on && !!node.response && !!node.response.transition
+  && !!getNextNodes(node);
+
+export const generateUniqueLabel = (label, takenLabels) => {
+  const usedLabels = takenLabels;
+
+  if (usedLabels[label]) {
+    usedLabels[label] += 1;
+    return `${label} (#${usedLabels[label]})`;
+  }
+
+  usedLabels[label] = 1;
+  return label;
+};
+
+export const getGraphWithUniqueLabels = (root, takenLabels = {}) => ({
+  ...root,
+  uniqueLabel: generateUniqueLabel(root.label, takenLabels),
+  ...(root.on
+    ? {
+      on: Object.entries(root.on)
+        .map(([transition, node]) => [transition, getGraphWithUniqueLabels(node, takenLabels)])
+        .reduce((total, [transition, node]) => ({ ...total, [transition]: node }), {}),
```

Comment:
https://www.richsnapp.com/blog/2019/06-09-reduce-spread-anti-pattern - I'll just leave it here.
tl;dr: reduce with spread operator has almost O(n^2) complexity.
I don't know if it's an issue for you.

Reply:
Good too know, I didn't realize that. But I think it's ok here, especially since an alternative would be to mutate a function parameter which is forbidden by our eslint and since most of the data will be fairly small so there won't be much difference anyway

Reply:
Sure, no problem 😄